### ISSUE TITLE
Stat Name Null

### ISSUE THREAD

**Issue by r2d2Proton** [Microsoft/xbox-live-api#216]
After getting the StatisticEventType::GetLeaderboardComplete event, the stat name contained in the LeaderboardRow is null:

                    LeaderboardResultEventArgs^ LbResultEventArgs = safe_cast<LeaderboardResultEventArgs^>(Event->EventArgs);
                    LeaderboardResult^ LbResult = LbResultEventArgs->Result;

                    for (LeaderboardRow^ Row : LbResult->Rows)
                    {
                        FString RowText = FString::Printf(TEXT("Leaderboard row: User %s, Stats "), Row->Gamertag->Data());
                        for (unsigned int i = 0; i < Row->Values->Size; ++i)
                        {
                            LeaderboardColumn^ statCol = LbResult->Columns->GetAt(i);
                            Platform::String^ statName = LbResult->Columns->GetAt(i)->DisplayName;

                            RowText.Append(FString::Printf(TEXT("(%s : %s) "), LbResult->Columns->GetAt(i)->DisplayName->Data(), Row->Values->GetAt(i)->Data()));
                        }

                        UE_LOG_ONLINE(Warning, *RowText);
                    }

**Comment by jasonsandlin** [comment]
Can you attach a fiddler trace?  https://docs.microsoft.com/en-us/windows/uwp/xbox-live/using-xbox-live/troubleshooting/how-to-set-up-fiddler-for-debugging

**Comment by r2d2Proton** [comment]
Is there another tool I can use?

**Comment by jasonsandlin** [comment]
You can route the HTTP calls to your own logging system:

https://docs.microsoft.com/en-us/windows/uwp/xbox-live/using-xbox-live/troubleshooting/troubleshooting-the-xbox-live-services-api

**Comment by r2d2Proton** [comment]
I am trying:
LiveContext->Settings->DiagnosticsTraceLevel = XboxServicesDiagnosticsTraceLevel::Verbose;

But the output in VS doesn't appear too helpful:
07/12/17 15:14:03 [19204] info  - Websocket connnection established.
07/12/17 15:14:03 [19204] debug  - websocket state change: connecting -> connected
07/12/17 15:14:03 [19204] debug  - Finish websocket connection task


Before that there are numerous Date Time entries in the log that are similar to those shown above, which leads me to believe they are relevant too. Would you like to see them all?

**Comment by jasonsandlin** [comment]
Try doing the EnableServiceCallRoutedEvents and ServiceCallRouted from that doc page.  It'll tell you what happened during the HTTP calls which you can log how you like

**Comment by r2d2Proton** [comment]
That didn't work as we hoped:
[2017.07.12-21.29.42:757][411]LogOnline:Warning: LIVE: [URL]:  
[2017.07.12-21.29.42:757][411]LogOnline:Warning: LIVE: [Response]:  

But there are many others above it that may be useful, for example:
07/12/17 15:29:42 [10540] info  - Presence service for the user already exsit, return
07/12/17 15:29:42 [11316] debug  - Token result properties:
07/12/17 15:29:42 [11316] debug  - Environment : prod
07/12/17 15:29:42 [11316] debug  - Gamertag : r2d2Electron
07/12/17 15:29:42 [11316] debug  - Restrictions : 195 229 230 234 251 252
07/12/17 15:29:42 [11316] debug  - Privileges : 188 191 192 193 194 196 198 199 200 201 203 204 205 206 207 208 211 214 215 216 217 220 224 227 228 235 238 245 247 249 254 255


Are these related?

**Comment by jasonsandlin** [comment]
No, you can turn off LiveContext->Settings->DiagnosticsTraceLevel as that isn't useful for this issue.
In your ServiceCallRouted lamda, what data is coming back?  Is it blank?

**Comment by r2d2Proton** [comment]
The code is as follows:
    LiveContext->Settings->DiagnosticsTraceLevel = XboxServicesDiagnosticsTraceLevel::Verbose;

    LiveContext->Settings->EnableServiceCallRoutedEvents = true;
    LiveContext->Settings->ServiceCallRouted += ref new
    Windows::Foundation::EventHandler<Microsoft::Xbox::Services::XboxServiceCallRoutedEventArgs^>(
    [=](Platform::Object^, Microsoft::Xbox::Services::XboxServiceCallRoutedEventArgs^ args)
    {
        UE_LOG_ONLINE(Warning, TEXT("[URL]: %s %s"), args->HttpMethod->ToString(), args->Url->AbsoluteUri->ToString());
        UE_LOG_ONLINE(Warning, TEXT("[Response]: %s %s"), args->HttpStatus.ToString(), args->ResponseBody->ToString());
    });

VS does break at the args->HttpMethod->ToString() line, and I can view args at a high level from the debugger. . .but I can't see the details.

**Comment by r2d2Proton** [comment]
Okay, a fix up helped:
[2017.07.12-21.50.03:683][475]LogOnline:Warning: LIVE: [URL]: GET https://sessiondirectory.xboxlive.com/
[2017.07.12-21.50.03:683][475]LogOnline:Warning: LIVE: [Response]: 200 {"results":[]}

I added the ->Data() call.

**Comment by jasonsandlin** [comment]
Ok cool.  Can you send me the logs for the entire run, include where you set stat and get the leaderboard.

**Comment by r2d2Proton** [comment]
Do you mean the VS log with XboxServicesDiagnosticsTraceLevel::Verbose?

If not, there isn't much else to show.

**Comment by jasonsandlin** [comment]
No, XboxServicesDiagnosticsTraceLevel isn't useful.  I need your "LogOnline:Warning: LIVE: [URL]:", etc logs for the full run (set the stat, get the leaderboard, etc) to see what's going on.

**Comment by r2d2Proton** [comment]
Can we work on this tomorrow? I need to take off now.

The flow is this, at startup, when logging in, I read the stats, which is what you see above. After playing a game, then I update the stats, which does take a little longer. I will get this for you later and post it here.

Thank you for your help.

**Comment by jasonsandlin** [comment]
Yep

**Comment by r2d2Proton** [comment]
The log, as it is, may not be very useful

Start Game:
[2017.07.13-02.01.54:690][  0]LogOnline:Warning: LIVE: No SecureDeviceAssociationTemplateName specified in the engine ini file.
[2017.07.13-02.01.54:690][  0]LogOnline:Warning: LIVE: FOnlineAchievementsLive: Failed to find json OSS achievements config: ../../../LongshotHero/Config/OSS/Live/Achievements.json

Login:
[2017.07.13-02.03.31:544][297]LogOnline:Warning: LIVE: [URL]: GET https://sessiondirectory.xboxlive.com/
[2017.07.13-02.03.31:544][297]LogOnline:Warning: LIVE: [Response]: 200 {"results":[]}
[2017.07.13-02.03.31:999][323]LogOnline:Warning: LIVE: DoWork LocalUserAdded: r2d2Electron
[2017.07.13-02.03.47:250][419]LogOnline:Warning: LIVE: DoWork GetLeaderboardComplete: r2d2Electron

Complete Level:
[2017.07.13-02.09.47:837][381]LogOnline:Warning: LIVE: DoWork StatisticUpdateComplete: r2d2Electron
[2017.07.13-02.09.47:838][381]LogOnline:Warning: LIVE: DoWork Stat: HighScore

Exit Game:
[2017.07.13-02.10.03:492][242]LogOnline:Warning: LIVE: FSessionMessageRouter::UnsubscribeFromMultiplayerEvents failed - couldn't get XboxLiveContext for user 2533274862546057.

When I perform a StatisticManager::SetStatisticIntegerData the StatisticUpdateComplete event triggers. How do I get the event results:

                case StatisticEventType::StatisticUpdateComplete:
                    UE_LOG_ONLINE(Warning, TEXT("DoWork StatisticUpdateComplete: %s"), Event->User->Gamertag->ToString()->Data());

                    IVectorView<Platform::String^>^ stats = mgr->GetStatisticNames(Event->User);
                    for (Platform::String^ stat : stats)
                    {
                        UE_LOG_ONLINE(Warning, TEXT("DoWork Stat: %s"), stat->ToString()->Data());
                    }

                    break;

**Comment by jasonsandlin** [comment]
It doesn't look like the stat value has been set on the server yet as there's no call going to statswrite.xboxlive.com.  

By default it'll send the stat changes every 5 mins.  If you need it to go out faster or if the app is closing, call RequestFlushToService:

https://github.com/Microsoft/xbox-live-api/blob/master/Source/Services/Stats/Manager/WinRT/StatisticManager_WinRT.h#L46

**Comment by r2d2Proton** [comment]
Actually, after I write the HighScore I call flush, otherwise I won't get the update event:
    Microsoft::Xbox::Services::Statistics::Manager::StatisticManager^ mgr = Microsoft::Xbox::Services::Statistics::Manager::StatisticManager::SingletonInstance;
    if (mgr != nullptr)
    {
        for (FStatPropertyArray::TConstIterator item(WriteObject.Properties); item; ++item)
        {
            Platform::String^ itemName = ref new Platform::String(*item->Key.GetPlainNameString());

            int32 itemValue;
            item->Value.GetValue(itemValue);

            long long llValue;
            llValue = itemValue;

            mgr->SetStatisticIntegerData(LiveContext->User, itemName, llValue);
        }

        mgr->RequestFlushToService(LiveContext->User);

**Comment by jasonsandlin** [comment]
Hmm, that RequestFlushToService should call ServiceCallRouted.  

If that's not happening, I recommend you set some break points inside XSAPI to see what's its doing. 
You can link to source and debug XSAPI easily but following this:

https://github.com/Microsoft/xbox-live-api/blob/master/LINKTOSOURCE.md#how-to-link-against-the-xsapi-winrt-uwp-source

Then set a breakpoints at to ensure each of these is getting called:

https://github.com/Microsoft/xbox-live-api/blob/master/Source/Services/Stats/Manager/WinRT/StatisticManager_WinRT.cpp#L68

https://github.com/Microsoft/xbox-live-api/blob/master/Source/Services/Stats/Manager/stats_manager_impl.cpp#L265

https://github.com/Microsoft/xbox-live-api/blob/master/Source/Services/Stats/Manager/stats_service.cpp#L33

**Comment by r2d2Proton** [comment]
Thanks Jason. Let me give it a try. I'll report back later. . .

**Comment by r2d2Proton** [comment]
Where can I get the 140 build tools?

Microsoft.Cpp.Platform.targets(57,5): error MSB8020: The build tools for v140 (Platform Toolset = 'v140') cannot be found. To build using the v140 build tools, please install v140 build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution".

**Comment by jasonsandlin** [comment]
From that doc page:

If you are using VS2017 for WinRT, we only have a single WinRT project that supports both VS2015 and VS2017. To use it from source, you'll need to install 140 support in VS2017. Run the VS2017 installer, modify, and go to Individual components tab and choose "VS++ 2015.3 v140 toolset". Then you should be able to open

**Comment by r2d2Proton** [comment]
Yes, I did that first.

**Comment by jasonsandlin** [comment]
Hmm, if the v140 toolset is installed, VS2017 should be able build it.   You can try targeting it to v141 if that doesn't work or install VS2015 otherwise

**Comment by r2d2Proton** [comment]
Converting them to v141 allowed the build to work. Does this affect the API/DLL at all? I mean in terms of any software dependent on it.

**Comment by jasonsandlin** [comment]
Nah, its fine.  We just have kept it at v140 so folks using VS2015 can build it.

**Comment by r2d2Proton** [comment]
Thanks. I'll debug with it now. . .

**Comment by r2d2Proton** [comment]
After updating to VS 15.2 I can no longer build my project. I'm getting the following error:

failed on 'C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcpackages\platform.winmd'
'The system cannot find the path specified.'

By chance do you know of a fix for this?

**Comment by jasonsandlin** [comment]
Hmm, that's an odd one.  I'm running the latest VS2017 and VS2015 and our samples don't hit that error.  I don't know how to fix that.  Maybe someone in https://social.msdn.microsoft.com/Forums/en-US/home?forum=wpdevelop would know?

**Comment by r2d2Proton** [comment]
Hey Jason, I am able to build my game (UE4 with UWP) now using VS 2017 ver. 15.2. I'm running into some problem with replacement of the REST and XSAPI DLL's since they are part of the build package. I'm trying to work that out now.

**Comment by jasonsandlin** [comment]
Since this issue is getting pretty hard to parse, I'm going to close this one out.  If you need more help, feel free to open a new issue or better yet post on MSDN at https://social.msdn.microsoft.com/Forums/en-US/home?forum=xboxlivedev.  Github issues list are best for bug reports while MSDN is best for I'm stuck questions so others who get stuck can easily search/find the topic.

**Comment by r2d2Proton** [comment]
Jason, I just got debugging working again. I will follow the steps mentioned in the link, along with your recommendation, and see if I can get any additional information. It is worth noting that my Stat Updates and Retrieval are not working.
